src/lib/services/RuleEngine.ts

```diff
@@ -7,7 +7,10 @@ export interface RuleEngine {
 
 	getCatalog(): Promise<Catalog>;
 
-	run(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[]): Promise<RuleResult[]>;
+	/**
+	 * @param engineOptions - a mapping of keys to values for engineOptions. not all key/value pairs will apply to all engines.
+	 */
+	run(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]>;
```

Comment:
How about making engineOptions an optional parameter? That way PmdEngine and EslintEngine don't even have to keep it in the method signature of this method's implementation.

Reply:
Also, we can default the value to empty map here so that not every caller has to remember to include an empty map if they don't have options to pass around.

Reply:
Something like this in the method signature:
`engineOptions?: Map<string, string> = new Map<string,string>()`

Reply:
I did a quick branch to try this out, you can see that diff here https://github.com/forcedotcom/sfdx-scanner/compare/d/jbartolotta/W-7760223-EngineOptionsOptional. I would need to write more unit tests if this was the route to go.

The question that I don't know the answer to in regards to Typescript best practices: `?` can be used so that the caller can optionally specify that argument, should `?` also be used so that the implementation of an interface can leave the argument out of the signature. 

I wasn't able to find anything on this question, my vote would be that it's clearer if the implementation's signature exactly matches the Interface signature.

Reply:
Code changes on the branch look nice.

I tried this with the EslintStrategy interface for `getTargetPatterns()` to optionally include `target` parameter - got away with excluding the parameter in JavascriptEslintStrategy's implementation of the same method. This also helped me avoid lint errors complaining about unused variables. 

That said, I haven't had any luck finding best practice suggestions for optional parameters either - if you feel more comfortable calling out all the parameters in the method signature, let's keep it as is.

Reply:
Ok, I'll commit this PR for 2.0 and we can revisit along with the other issues with the `run` signature in the quip doc for 3.0.